cluster-autoscaler/cloudprovider/aws/README.md

```diff
@@ -0,0 +1,79 @@
+# Cluster Autoscaler on AWS 
+The cluster autoscaler on AWS scales worker nodes within an autoscaling group. It will run as a `Deployment` in your cluster. This README will go over some of the necessary steps required to get the cluster autoscaler up and running. 
+
+## Kubernetes Version 
+Cluster autoscaler must run on v1.3.0 or greater. 
+
+## Permissions 
+The worker running the cluster autoscaler will need access to certain resources and actions:
+```json
+{
+    "Version": "2012-10-17",
+    "Statement": [
+        {
+            "Effect": "Allow",
+            "Action": [
+                "autoscaling:DescribeAutoScalingGroups",
+                "autoscaling:DescribeAutoScalingInstances",
+                "autoscaling:SetDesiredCapacity",
+                "autoscaling:TerminateInstanceInAutoScalingGroup"
+            ],
+            "Resource": "*"
+        }
+    ]
+}
+```
+Unfortunately AWS does not support ARNs for autoscaling groups yet so you must use "*" as the resource.
+
+## Deployment Specification
+Your deployment configuration should look something like this:
+```yaml
+---
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: cluster-autoscaler
+  labels:
+    app: cluster-autoscaler
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: cluster-autoscaler
+  template:
+    metadata:
+      labels:
+        app: cluster-autoscaler
+    spec:
+      containers:
+        - image: {{ YOUR IMAGE HERE }}
+          name: cluster-autoscaler
+          resources:
+            limits:
+              cpu: 100m
+              memory: 300Mi
+            requests:
+              cpu: 100m
+              memory: 300Mi
+          command:
+            - ./cluster-autoscaler
+            - -v=4
+            - --cloud-provider=aws
+            - --skip-nodes-with-local-storage=false
+            - --nodes={{ ASG MIN e.g. 1 }}:{{ASG MAX e.g. 5}}:{{ASG NAME e.g. k8s-worker-asg}}
+          env:
+            - name: AWS_REGION
+              value: us-east-1
+          volumeMounts:
+            - name: ssl-certs
+              mountPath: /etc/ssl/certs/ca-certificates.crt
+              readOnly: true
+          imagePullPolicy: "Always"
+      volumes:
+        - name: ssl-certs
+          hostPath:
+            path: "/etc/ssl/certs/ca-certificates.crt"
+```
+Note: 
+- the `/etc/ssl/certs/ca-certificates.crt` should exist by default on your ec2 instance.
+- at the time of writing this, cluster autoscaler is unaware of availability zones, the availability zone of the instance should be configured by the autoscaling group. Although autoscaling groups can contain instances in multiple availability zones, the autoscaling group should span 1 availability zone for the cluster autoscaler to work.  
```

Comment:
@mumoshu, it will likely work with a multi-AZ ASG, but there may be some caveats if Kubernetes is configured to be zone-aware.  Pardon the long-winded explanation.  Here is the reasoning:

The cluster-autoscaler asks the AWS CloudProvider for a sample Node from the NodeGroup (backed by an ASG), and uses it to make scaling decisions.  It assumes that the sample Node is equivalent to all other nodes in the ASG - eg. same instance type, storage, etc.  When it needs to scale up, for example, it will know with certainty that new Nodes will have the same capacity and will be able to accommodate the pending pods.

The cluster-autoscaler has logic that simulates the Scheduler's decisions to see if a new Node will be able to accommodate pending workloads.  If the Scheduler is zone-aware, it may specifically want to distribute workloads across AZs.

Consider this scenario:

* The Scheduler wants to run some Pods in Zone A because there are already equivalent pods in Zone B and it wants a multi-AZ Pod distribution.
* There are not enough Nodes in Zone A
* In this situation the Pods will be pending, waiting for a Node in Zone A
* The cluster autoscaler sees pending pods and starts looking for a NodeGroup (an ASG) that can accommodate these Pods
* Say we have a NodeGroup whose ASG spans Zones A and B
* This NodeGroup returns a random sampleNode that is in Zone A
* The cluster-autoscaler says "Great, this NodeGroup has an appropriate node, let me scale it up", and it increases the DesiredCapacity by 1
* The ASG launches a Node in Zone B (because it is trying to keep the Zones balanced)
* The new Node appears in Kubernetes and the Scheduler sees that is in Zone B
* The Scheduler continues to wait a Node in Zone A in order to schedule the pending Pods
* The new Node did not help accommodate the pending Pods

Eventually the cluster-autoscaler will try to launch another Node because there are still Pods pending.  By chance if a new Node lands in Zone A, the Pods will get scheduled and things will work, but this would be non-deterministic and would make the process less reliable

The crux of the matter is the contract between NodeGroup that provides a sampleNode, and the scaling logic, which expects that a sampleNode is equivalent to all other nodes in that NodeGroup.

These are theoretical, I haven't tested these scenarios, but from my understanding of the Node sampling logic, something like this would happen.